src/linear_algebra/matrix/reindex.lean

```diff
@@ -31,112 +32,120 @@ open_locale matrix
+lemma reindex_linear_equiv_comp (e₁ : m ≃ m') (e₂ : n ≃ n') (e₁' : m' ≃ m'')
   (e₂' : n' ≃ n'') :
-  (reindex_linear_equiv e₁' e₂' : _ ≃ₗ[R] _) ∘ (reindex_linear_equiv e₁ e₂ : _ ≃ₗ[R] _)
-  = reindex_linear_equiv (e₁.trans e₁') (e₂.trans e₂') :=
+  (reindex_linear_equiv R A e₁' e₂') ∘ (reindex_linear_equiv R A e₁ e₂)
+  = reindex_linear_equiv R A (e₁.trans e₁') (e₂.trans e₂') :=
 by { rw [← reindex_linear_equiv_trans], refl }
 
-lemma reindex_linear_equiv_comp_apply [semiring R] (e₁ : m ≃ m') (e₂ : n ≃ n') (e₁' : m' ≃ m'')
-  (e₂' : n' ≃ n'') (M : matrix m n R) :
-  (reindex_linear_equiv e₁' e₂') (reindex_linear_equiv e₁ e₂ M) =
-    reindex_linear_equiv (e₁.trans e₁') (e₂.trans e₂') M :=
+lemma reindex_linear_equiv_comp_apply (e₁ : m ≃ m') (e₂ : n ≃ n') (e₁' : m' ≃ m'')
+  (e₂' : n' ≃ n'') (M : matrix m n A) :
+  (reindex_linear_equiv R A e₁' e₂') (reindex_linear_equiv R A e₁ e₂ M) =
+    reindex_linear_equiv R A (e₁.trans e₁') (e₂.trans e₂') M :=
 minor_minor _ _ _ _ _
 
-lemma reindex_linear_equiv_one [semiring R] [decidable_eq m] [decidable_eq m']
-  (e : m ≃ m') : (reindex_linear_equiv e e (1 : matrix m m R)) = 1 :=
+lemma reindex_linear_equiv_one [decidable_eq m] [decidable_eq m'] [has_one A]
+  (e : m ≃ m') : (reindex_linear_equiv R A e e (1 : matrix m m A)) = 1 :=
 minor_one_equiv e.symm
 
-variables {o o' : Type*} [fintype o] [fintype o']
+end add_comm_monoid
+
+section semiring
+variables [semiring R] [semiring A] [module R A]
 
-lemma reindex_linear_equiv_mul [semiring R]
-  (eₘ : m ≃ m') (eₙ : n ≃ n') (eₒ : o ≃ o') (M : matrix m n R) (N : matrix n o R) :
-  reindex_linear_equiv eₘ eₒ (M ⬝ N) =
-    reindex_linear_equiv eₘ eₙ M ⬝ reindex_linear_equiv eₙ eₒ N :=
+lemma reindex_linear_equiv_mul
+  (eₘ : m ≃ m') (eₙ : n ≃ n') (eₒ : o ≃ o') (M : matrix m n A) (N : matrix n o A) :
+  reindex_linear_equiv R A eₘ eₒ (M ⬝ N) =
+    reindex_linear_equiv R A eₘ eₙ M ⬝ reindex_linear_equiv R A eₙ eₒ N :=
 minor_mul_equiv M N _ _ _
 
-lemma mul_reindex_linear_equiv_one [semiring R] [decidable_eq o] (e₁ : o ≃ n) (e₂ : o ≃ n')
-  (M : matrix m n R) : M.mul (reindex_linear_equiv e₁ e₂ 1) =
-    reindex_linear_equiv (equiv.refl m) (e₁.symm.trans e₂) M :=
+lemma mul_reindex_linear_equiv_one [decidable_eq o] (e₁ : o ≃ n) (e₂ : o ≃ n')
+  (M : matrix m n A) : M.mul (reindex_linear_equiv R A e₁ e₂ 1) =
+    reindex_linear_equiv R A (equiv.refl m) (e₁.symm.trans e₂) M :=
 mul_minor_one _ _ _
 
+end semiring
+
+section algebra
+
+variables [comm_semiring R] [decidable_eq m] [decidable_eq n]
+
 /--
 For square matrices with coefficients in commutative semirings, the natural map that reindexes
 a matrix's rows and columns with equivalent types, `matrix.reindex`, is an equivalence of algebras.
 -/
-def reindex_alg_equiv [comm_semiring R] [decidable_eq m] [decidable_eq n]
-  (e : m ≃ n) : matrix m m R ≃ₐ[R] matrix n n R :=
+def reindex_alg_equiv (e : m ≃ n) : matrix m m R ≃ₐ[R] matrix n n R :=
```

Comment:
Obviously it would be nice to have `algebra R A` imply 
```suggestion
def reindex_alg_equiv (e : m ≃ n) : matrix m m A ≃ₐ[R] matrix n n A :=
```
but we don't have the corresponding `algebra R (matrix n n A)` structure needed to declare it. I might follow up with that in another PR.